endpoints/events/vtrack.go

```diff
@@ -0,0 +1,306 @@
+
+	// handle pbs caching errors
+	if len(errs) != 0 {
+		glog.Errorf("Error(s) updating vast: %v", errs)
+		return nil, errs
+	}
+
+	// build response
+	response := &BidCacheResponse{
+		Responses: []CacheObject{},
+	}
+
+	for _, uuid := range r {
+		response.Responses = append(response.Responses, CacheObject{
+			UUID: uuid,
+		})
+	}
+
+	return response, nil
+}
+
+// Cache BidCacheRequest data
+func (v *vtrackEndpoint) cachePutObjects(req *BidCacheRequest, biddersAllowingVastUpdate []string, accountId string, timeout int64) ([]string, []error) {
+	var cacheables []prebid_cache_client.Cacheable
+	sort.Strings(biddersAllowingVastUpdate)
+
+	for _, c := range req.Puts {
+
+		nc := &prebid_cache_client.Cacheable{
+			Type:       c.Type,
+			Data:       c.Data,
+			TTLSeconds: c.TTLSeconds,
+			Key:        c.Key,
+		}
+
+		if contains(biddersAllowingVastUpdate, c.Bidder) && nc.Data != nil {
+			nc.Data = modifyVastXml(v.Cfg.ExternalURL, nc.Data, c.BidID, c.Bidder, accountId, c.Timestamp)
+		}
+
+		cacheables = append(cacheables, *nc)
+	}
+
+	t := time.Now().Add(time.Duration(timeout) * time.Millisecond)
+	ctx, cancel := context.WithDeadline(context.Background(), t)
+	defer cancel()
+
+	return v.Cache.PutJson(ctx, cacheables)
+}
+
+// Returns list of bidders that allow VAST XML modification.
+func biddersAllowingVastUpdate(req *BidCacheRequest, bidderInfos *adapters.BidderInfos, allowUnknownBidder bool) []string {
+	var bl = make(map[string]struct{})
+
+	for _, bcr := range req.Puts {
+		if _, ok := bl[bcr.Bidder]; isAllowVastForBidder(&bcr, bidderInfos, allowUnknownBidder) && !ok {
+			bl[bcr.Bidder] = struct{}{}
+		}
+	}
+
+	length := len(bl)
+	if length == 0 {
+		return []string{}
+	}
+
+	keys := make([]string, 0, length)
+	for key := range bl {
+		keys = append(keys, key)
+	}
+
+	return keys
+}
+
+// Checks if Bidder is active and allowed to modify vast xml data
+func isAllowVastForBidder(r *prebid_cache_client.Cacheable, bidderInfos *adapters.BidderInfos, allowUnknownBidder bool) bool {
+	//if bidder is active and isModifyingVastXmlAllowed is true
+	// check if bidder is configured
+	if b, ok := (*bidderInfos)[r.Bidder]; bidderInfos != nil && ok {
+		// check if bidder is enabled
+		return b.Status == adapters.StatusActive && b.ModifyingVastXmlAllowed
+	}
+
+	return allowUnknownBidder
+}
+
+// Extracts account id from a HTTP Request
+func getAccountId(httpRequest *http.Request) string {
+	return httpRequest.FormValue(AccountParameter)
+}
+
+// Modify BidCacheRequest element Vast XML data
+func modifyVastXml(externalUrl string, data json.RawMessage, bidid string, bidder string, accountId string, timestamp int64) json.RawMessage {
+	c := string(data)
+	ci := strings.Index(c, ImpressionCloseTag)
+
+	// no impression tag - pass it as it is
+	if ci == -1 {
+		return data
+	}
+
+	vastUrlTracking := GetVastUrlTracking(externalUrl, bidid, bidder, accountId, timestamp)
```

Comment:
Please add test coverage from line 291 and down. The current tests always return from the `ci == -1` check.

Reply:
There is still no coverage. I think the issue is with the test setup. Please look for my comment there.